src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/impl/InternalEvaluatorImpl.kt

```diff
@@ -124,4 +146,69 @@ internal class InternalEvaluatorImpl(
             isExecuting = false
         }
     }
+
+    private fun updateVariablesState(cellId: Int) {
+        // remove known modifying usages in this cell
+        varsUsagePerCell[cellId]?.removeIf {
+            varsDeclarationInfo[it] != cellId
+        }
+
+        variablesHolder.forEach {
+            val state = it.value
+            val oldValue = state.stringValue
+            state.update()
+
+            if (state.stringValue != oldValue) {
+                varsUsagePerCell[cellId]?.add(it.key)
```

Comment:
So, we store modifications here.. Why is it "usage" map then? And how do we use this map, why is it needed for?

Reply:
"Usage" was motivated by the means that we stored vars that were used in modifying behaviour. We might use it to output only "interesting" vars per cell to the user